packages/compiler-cli/src/ngtsc/transform/src/transform.ts

```diff
@@ -43,69 +44,94 @@ export function ivyTransformFactory(
-            field.initializer, this.importManager, this.defaultImportRecorder,
-            ts.ScriptTarget.ES2015);
-
-        // Create a static property declaration for the new field.
-        const property = ts.createProperty(
-            undefined, [ts.createToken(ts.SyntaxKind.StaticKeyword)], field.name, undefined,
-            undefined, exprNode);
-
-        if (this.isClosureCompilerEnabled) {
-          // Closure compiler transforms the form `Service.ɵprov = X` into `Service$ɵprov = X`. To
-          // prevent this transformation, such assignments need to be annotated with @nocollapse.
-          // Note that tsickle is typically responsible for adding such annotations, however it
-          // doesn't yet handle synthetic fields added during other transformations.
-          ts.addSyntheticLeadingComment(
-              property, ts.SyntaxKind.MultiLineCommentTrivia, '* @nocollapse ',
-              /* hasTrailingNewLine */ false);
-        }
-
-        field.statements
-            .map(
-                stmt => translateStatement(
-                    stmt, this.importManager, this.defaultImportRecorder, ts.ScriptTarget.ES2015))
-            .forEach(stmt => statements.push(stmt));
-
-        members.push(property);
-      });
-
-      // Replace the class declaration with an updated version.
-      node = ts.updateClassDeclaration(
-          node,
-          // Remove the decorator which triggered this compilation, leaving the others alone.
-          maybeFilterDecorator(node.decorators, this.compilation.decoratorsFor(node)),
-          node.modifiers, node.name, node.typeParameters, node.heritageClauses || [],
-          // Map over the class members and remove any Angular decorators from them.
-          members.map(member => this._stripAngularDecorators(member)));
-      return {node, after: statements};
+    const result = this.compilation.compile(node, this.constantPool);
+    if (result !== null) {
+      this.classCompilationMap.set(node, result);
     }
-
     return {node};
   }
+}
+
+/**
+ * Visits all classes and performs transformation of corresponding TS nodes based on the Ivy
+ * compilation results (provided as an argument).
+ */
+class IvyTransformationVisitor extends Visitor {
+  constructor(
+      private compilation: TraitCompiler,
+      private classCompilationMap: Map<ts.ClassDeclaration, CompileResult[]>,
+      private reflector: ReflectionHost, private importManager: ImportManager,
+      private defaultImportRecorder: DefaultImportRecorder,
+      private isClosureCompilerEnabled: boolean, private isCore: boolean) {
+    super();
+  }
+
+  visitClassDeclaration(node: ts.ClassDeclaration):
+      VisitListEntryResult<ts.Statement, ts.ClassDeclaration> {
+    // If this class is not registered in the map, it means that it doesn't have Angular decorators,
+    // thus no further processing is required.
+    if (!this.classCompilationMap.has(node)) return {node};
```

Comment:
Style in the compiler is to always use braces:
```suggestion
    if (!this.classCompilationMap.has(node)) {
      return {node};
    }
```